### PR TITLE
[Refactor] Disagg Gemma4: models/demos/gemma4_d_p

### PR DESCRIPTION
**Pull request by svuckovicTT**
### PR Category
Cleanup

### Summary
Simplify the Gemma4 prefill implementation and make ownership and configuration explicit.

The disaggregated prefiil implementation was initially written in `models/demos/gemma4` and then moved to a standalone dir: `models/demos/gemma4_d_p`. This PR sheds unnecessary fat that was used to support multiple HW impls and all gemma4 variants, and many other things, greatly simplifying the codeflow and improving readability.

- **Configuration:** bind the device to `MeshConfig`, standardize CP/TP naming, remove unused mode/expert settings, and require precision and chunk-size arguments.
- **Weight loading:** use multimodal checkpoint keys, remove redundant guards and casts, skip LM head and final norm, and build only the required QK or QKV projection.
- **Attention and caches:** inline forwarding wrappers, standardize global/sliding function names, use named cache containers, and explicitly release temporary cast tensors.
- **Request metadata:** move slot and KV-position tensors from `CCLManager` into model-owned `PrefillMetadata`, preserving stable addresses for trace replay.
- **Trace and memory handling:** remove host semaphore resets, correct async all-gather semaphore counts.
- **Readability:** simplify conditionals, argument handling, variable names, comments, and input staging.

### Notes for reviewers
Focus on metadata propagation, tensor lifetimes, and trace replay behavior. Final model output is now the last decoder layer’s hidden states, without final normalization.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:svuckovic/gemma4-readability-improvements-pr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:svuckovic/gemma4-readability-improvements-pr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:svuckovic/gemma4-readability-improvements-pr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:svuckovic/gemma4-readability-improvements-pr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:svuckovic/gemma4-readability-improvements-pr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:svuckovic/gemma4-readability-improvements-pr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:svuckovic/gemma4-readability-improvements-pr)

### CODE REVIEW

File: models/demos/gemma4_d_p/demo/text_demo_prefill.py [hunk 1]
```diff
@@ -233,85 +206,74 @@ def test_prefill_long_context_traced(
...
             tokens_all[:, chunk_start : chunk_start + chunk_size].contiguous(),
-            ttnn.uint32,
-            ttnn.ROW_MAJOR_LAYOUT,
-            mesh_config=mesh_config,
-            seq_dim=-1,
+            device=None,
+            dtype=ttnn.uint32,
+            layout=ttnn.ROW_MAJOR_LAYOUT,
+            mesh_mapper=_cp_or_replicate_mapper(mesh_config, seq_dim=-1),
         )
```
[the remark this answers is not in the document]
Comment (svuckovicTT) [reply models/demos/gemma4_d_p/demo/text_demo_prefill.py:253]: Bug was fixed in SDPA already, reset logic was stale.

File: models/demos/gemma4_d_p/tt/model.py [hunk 2]
```diff
@@ -169,36 +167,31 @@ class Gemma4Model:
...
         num_layers=None,
-        precision=None,
-        # Global prefill chunk size. Only needed under context parallelism with more
-        # than one chunk: it sets the RoPE cache's chunk-major row order and sizes the
-        # ring KV cache slabs. None means single-chunk prefill.
-        prefill_chunk_size=None,
         ring_kv_caches=None,
     ):
-        from models.demos.gemma4_d_p.config import validate_galaxy_mesh
+        assert state_dict and any(
```
[the remark this answers is not in the document]
Comment (svuckovicTT) [reply models/demos/gemma4_d_p/tt/model.py:147]: Only the multi-modal path succeeds, and we don't support text-only path, and don't plan to, hence the change.

File: models/demos/gemma4_d_p/tt/model.py [hunk 3]
```diff
@@ -438,18 +373,17 @@ def __call__(
                 else:
                     ttnn.synchronize_device(self.mesh_device)
                     on_layer_complete(i)
-        return self.norm.forward(hidden_states)
+        return hidden_states
```
[the remark this answers is not in the document]
Comment (svuckovicTT) [reply models/demos/gemma4_d_p/tt/model.py:338]: There are no users that read the output logits. This change is intended.

File: models/demos/gemma4_d_p/tt/ccl.py [hunk 4]
```diff
@@ -10,31 +10,8 @@
...
-    # replay deadlock involves the two parallel link workers racing each other.
-    import os as _os
-
-    _forced = _os.environ.get("GEMMA4_NUM_LINKS")
-    if _forced:
-        return int(_forced)
-    return _default_num_links_impl()
-
-
 def _default_num_links_impl():
```
[the remark this answers is not in the document]
Comment (svuckovicTT) [reply models/demos/gemma4_d_p/tt/ccl.py:13]: Intentional.

File: models/demos/gemma4_d_p/tt/model_config.py [hunk 5]
```diff
@@ -256,15 +216,20 @@ def resolve_model_cache_path(model_path):
...
             raise ValueError("model_cache_path must be initialized before requesting a weight cache path")
         dtype_str = dtype_to_str(dtype)
         shape = mesh_shape if mesh_shape is not None else getattr(self, "cluster_shape", None)
         if shape is None:
             raise ValueError("Mesh shape must be initialized before requesting a weight cache path")
         mesh_suffix = "x".join(str(d) for d in shape)
-        return _ensure_cache_dir(self.model_cache_path / f"tensor_cache_{dtype_str}_mesh{mesh_suffix}")
+        path = self.model_cache_path / f"tensor_cache_{dtype_str}_mesh{mesh_suffix}"
+        if not path.is_dir():
+            raise FileNotFoundError(f"Weight cache directory does not exist or is not a directory: {path}")
```
[the remark this answers is not in the document]
Comment (svuckovicTT) [reply models/demos/gemma4_d_p/tt/model_config.py:234]: This is intentional.

File: models/demos/gemma4_d_p/tt/precision.py [hunk 6]
```diff
@@ -65,25 +64,14 @@ def load(cls, model_path, mesh_shape):
...
-            model_key = hub_match.group(1)
-        mesh_key = f"{mesh_shape[0]}x{mesh_shape[1]}"
-
         try:
             with open(_PATH) as f:
                 table = json.load(f)
         except FileNotFoundError:
             return cls({})
 
+        model_key = os.path.basename(str(model_path).rstrip("/"))
```
[the remark this answers is not in the document]
Comment (svuckovicTT) [reply models/demos/gemma4_d_p/tt/precision.py:75]: Making the miss explicit with an error message. No default (backup) path supported.

File: models/demos/gemma4_d_p/tt/model_config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (svuckovicTT) [reply models/demos/gemma4_d_p/tt/model_config.py:234]: This is intentional. Weights should be read off the cache mount. SIlently falling back to downloading and reserializing weights has its own risks.

**Review by sdjordjevicTT — APPROVED** [review]
